compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp

```diff
@@ -0,0 +1,220 @@
+
+  Location loc = linalgOp.getLoc();
+  int64_t mPadding = mnkPadding[0];
+  int64_t nPadding = mnkPadding[1];
+  int64_t kPadding = mnkPadding[2];
+  if (mPadding != 0 || kPadding != 0) {
+    // For NHWC, the m-padding is for W and k-padding is for C
+    newInput =
+        getPaddedValue(rewriter, loc, newInput, {0, 0, mPadding, kPadding});
+  }
+  if (nPadding != 0 || kPadding != 0) {
+    // For HWCF, the n-padding is for F and k-padding is for C
+    newFilter =
+        getPaddedValue(rewriter, loc, newFilter, {0, 0, kPadding, nPadding});
+  }
+  if (mPadding != 0 || nPadding != 0) {
+    // For output, the m-padding is for W and k-padding is for F
+    newOuts =
+        getPaddedValue(rewriter, loc, newOuts, {0, 0, mPadding, nPadding});
+  }
+
+  linalg::LinalgOp paddedConv2dOp =
+      mlir::clone(rewriter, linalgOp, {newOuts.getType()},
+                  ArrayRef<Value>{newInput, newFilter, newOuts});
+  // Extract slice.
+  IntegerAttr zero = rewriter.getI64IntegerAttr(0);
+  IntegerAttr one = rewriter.getI64IntegerAttr(1);
+  SmallVector<OpFoldResult> offsets(4, zero);
+  SmallVector<OpFoldResult> strides(4, one);
+  auto resultType = cast<RankedTensorType>(linalgOp->getResult(0).getType());
+  ArrayRef<int64_t> resultShape = resultType.getShape();
+  SmallVector<OpFoldResult> sizes = {rewriter.getIndexAttr(resultShape[0]),
+                                     rewriter.getIndexAttr(resultShape[1]),
+                                     rewriter.getIndexAttr(resultShape[2]),
+                                     rewriter.getIndexAttr(resultShape[3])};
+  Value extracted = rewriter.createOrFold<tensor::ExtractSliceOp>(
+      loc, paddedConv2dOp->getResults()[0], offsets, sizes, strides);
+  rewriter.replaceOp(linalgOp, extracted);
+}
+
+struct PadToIntrinsicsPass final
+    : impl::PadToIntrinsicsBase<PadToIntrinsicsPass> {
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    Operation *rootOp = getOperation();
+    ArrayAttr mmaKinds = nullptr;
+    for (IREE::HAL::ExecutableTargetAttr targetAttr :
+         IREE::HAL::DeviceTargetAttr::lookupExecutableTargets(rootOp)) {
+      FailureOr<ArrayAttr> candidateMmaKinds =
+          getSupportedMmaTypes(targetAttr.getConfiguration());
+      if (succeeded(candidateMmaKinds)) {
+        mmaKinds = *candidateMmaKinds;
+        break;
+      }
+    }
+    if (!mmaKinds)
+      return;
+
+    auto intrinsics = llvm::map_to_vector(
+        mmaKinds.getAsRange<IREE::GPU::MMAAttr>(), [](IREE::GPU::MMAAttr mma) {
+          auto [mSize, nSize, kSize] = mma.getMNKShape();
+          auto [aType, bType, cType] = mma.getABCElementTypes();
+          return GPUMatmulShapeType{mSize, nSize, kSize, aType, bType, cType};
+        });
+
+    SmallVector<linalg::LinalgOp> targetOps;
+    rootOp->walk([&](linalg::LinalgOp linalgOp) {
+      if (isa<linalg::Conv2DNhwcHwcfOp, linalg::BatchMatmulOp>(
```

Comment:
NIT: Maybe we can drop the BMMOp for now.

Reply:
Good catch, thanks